File: Python/problem0557.py
# -*- coding: utf-8 -*-
"""
Created on Sat Jul 20 15:51:52 2019

@author: ASUS
"""

class Solution:
    def reverseWords(self, s: str) -> str:
#        from collections import deque
#        s = deque(s)
        s = list(s)
#        s.reverse()
        stack = []
        ans = []
        while s:
            tmp = s.pop()
            if tmp != ' ':
                stack.append(tmp)
            else:
                ans.append(''.join(stack))
                stack = []
        ans.append(''.join(stack))
        ans.reverse()
        return ' '.join(ans)
        
        
solu = Solution()
s = "Let's take LeetCode contest"
print(solu.reverseWords(s))

File: C++/problem0557.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
#include <cmath>
#include <ctime>
#include <cstdlib>
#include <windows.h>

using namespace std;

class Solution {
public:
    string reverseWords(string s) {
        string stack, ans;
        // cout << boolalpha;
        while (s.empty() == false)
        // while (!s.empty())
        {
            // char tmp = s[s.size()-1];
            char tmp = *(s.end()-1);
            s.pop_back();
            if (tmp != ' ')
            {
                stack.push_back(tmp);
            }
            else
            {
                ans.insert(ans.begin(), stack.begin(), stack.end());
                string temp = " ";
                ans.insert(ans.begin(), temp.begin(), temp.end());
                stack = "";
            }
        }
        ans.insert(ans.begin(), stack.begin(), stack.end());
        return ans;
    }
};

int main()
{
    string s = "Let's take LeetCode contest";
    Solution solu;
    string ans = solu.reverseWords(s);
    cout << ans << endl;
    return 0;
}